### PR TITLE
Refactored PackageDatabase into ROS1 and ROS2 variants that implement a common interface

### PR DESCRIPTION
**Pull request by schmerl**

### CODE REVIEW

**Comment by schmerl** [comment]
After this, these issues should be resolved: #408, #406, #416

File: src/roswire/common/package.py [hunk 1]
```diff
@@ -204,7 +159,7 @@ def _from_paths(cls,
                     raise
             else:
                 packages.append(package)
-        return PackageDatabase(packages, paths)
+        return cls(packages, paths)
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:162]: This assumes knowledge of the constructor for `PackageDatabase` subclasses. We should probably add and use a abstract class method `for_packages_and_paths(packages, paths)` here instead.

File: src/roswire/common/package.py [hunk 2]
```diff
@@ -237,9 +192,14 @@ def __iter__(self) -> Iterator[str]:
         """
         yield from self.__contents
 
-    @staticmethod
-    def from_dict(d: List[Dict[str, Any]]) -> 'PackageDatabase':
-        return PackageDatabase((Package.from_dict(dd) for dd in d), [])
+    @classmethod
+    def _from_dict_internal(cls, d: List[Dict[str, Any]]) -> 'PackageDatabase':
+        return cls([Package.from_dict(dd) for dd in d], [])
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:197]: Same as above. Assumes implementation knowledge of the constructor.  `for_packages_and_paths` can do the job here.`

File: src/roswire/ros1/package.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+class ROS1PackageDatabase(PackageDatabase):
+
+    @classmethod
+    def from_dict(cls, d: List[Dict[str, Any]]) -> 'PackageDatabase':
+        return cls._from_dict_internal(d)
+
+    @classmethod
+    def _determine_paths(cls, app_instance: 'AppInstance') -> List[str]:
+
```
Comment (ChrisTimperley) [inline src/roswire/ros1/package.py:25]: excess space

File: src/roswire/ros1/package.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+from loguru import logger
+
+from ..common import PackageDatabase
+
+if typing.TYPE_CHECKING:
+    from .. import AppInstance
+
+
+class ROS1PackageDatabase(PackageDatabase):
+
```
Comment (ChrisTimperley) [inline src/roswire/ros1/package.py:18]: excess space

File: src/roswire/ros2/package.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
+__all__ = ('ROS2PackageDatabase',)
```
Comment (ChrisTimperley) [inline src/roswire/ros2/package.py:2]: add empty line after `__all__`

File: src/roswire/common/package.py [hunk 6]
```diff
@@ -122,51 +112,16 @@ def build(cls,
...
-        package_to_prefix: Mapping[str, str] = json.loads(jsn)
-        paths: List[str] = [os.path.join(prefix, f'share/{package}')
-                            for (package, prefix) in package_to_prefix.items()]
-        return paths
-
-    @classmethod
+    @abstractmethod
     def _determine_paths(cls, app_instance: 'AppInstance') -> List[str]:
-        """Parses :code:`ROS_PACKAGE_PATH` for a given shell.
-
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:160]: Missing empty line in docstring

File: src/roswire/common/package.py [hunk 7]
```diff
@@ -204,7 +182,7 @@ def _from_paths(cls,
                     raise
             else:
                 packages.append(package)
-        return PackageDatabase(packages, paths)
+        return cls._from_pacakages_and_paths(packages, paths)
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:185]: typo in method name

File: src/roswire/common/package.py [hunk 8]
```diff
@@ -122,51 +112,39 @@ def build(cls,
...
-        return cls._paths_ros1(app_instance)
+        ...
+
+    @classmethod
+    @abstractmethod
+    def _from_pacakages_and_paths(cls,
+                                  packages: Iterable[Package],
+                                  paths: Iterable[str]) -> 'PackageDatabase':
+        """
+        Constructs a packaged database from a packages
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:133]: Try to shift this onto the same line as `"""` if it will fit.

File: src/roswire/common/package.py [hunk 9]
```diff
@@ -122,51 +112,39 @@ def build(cls,
...
             list of paths should be obtained
         """
-        if app_instance.description.distribution.ros == ROSVersion.ROS2:
-            return cls._paths_ros2(app_instance)
-        return cls._paths_ros1(app_instance)
+        ...
+
+    @classmethod
+    @abstractmethod
+    def _from_pacakages_and_paths(cls,
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:129]: typo in method name

File: src/roswire/common/package.py [hunk 10]
```diff
@@ -122,51 +112,39 @@ def build(cls,
...
+        Parameters
+        ----------
+        packages: Iterable[Package]
+            A collection of the packages to be included in the databse
+        paths: Iterable[str]
+            A collection of paths
+
+        Returns
+        -------
+        A Package database
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:145]: I think we can drop the Returns section of this docstring.

File: test/test_format.py [hunk 11]
```diff
@@ -5,8 +5,10 @@
 import os
 
 from roswire.common import (Constant, Field, MsgFormat, SrvFormat,
-                            ActionFormat, PackageDatabase,
-                            FormatDatabase)
+                                 ActionFormat, Time, Package, PackageDatabase,
```
Comment (ChrisTimperley) [inline test/test_format.py:8]: Indentation looks a little borked on this

File: src/roswire/common/package.py [hunk 12]
```diff
@@ -122,51 +112,39 @@ def build(cls,
...
-                            for (package, prefix) in package_to_prefix.items()]
-        return paths
-
-    @classmethod
+    @abstractmethod
     def _determine_paths(cls, app_instance: 'AppInstance') -> List[str]:
-        """Parses :code:`ROS_PACKAGE_PATH` for a given shell.
-
+        """
+        Parses the package paths for a given shell.
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:118]: Ah. Sorry for the ambiguity. This text should go on the same line as the `"""`. We need the blank line between the method description and the Parameters section header.

**Review by ChrisTimperley — APPROVED** [review]
👍🏻 LGTM

**Pull request merged**